src/message-stream.ts

```diff
@@ -301,8 +298,14 @@ export class MessageStream extends PassThrough {
    */
   private _onEnd(stream: PullStream, status: StatusObject): void {
     this._removeStream(stream);
-
-    if (RETRY_CODES.includes(status.code)) {
+    if (Date.now() - this._timeSinceFirstError > this._options.timeout!) {
+      this._retries = 0;
+      this._timeSinceFirstError = Date.now();
+    }
+    if (
+      RETRY_CODES.includes(status.code) &&
+      this._retries++ < this._options.maxRetries!
```

Comment:
It isn't really uncommon for all the streams to end around the same time, I think if this were to happen then we would likely blow past the default `maxRetries` pretty quick and end up having too few streams open at any given time.

Reply:
_fillStreamPool actually fills up stream pool, so it will create all the stream it requires to fill up the pool back even on last retry. The scenario you mentioned can happen only if there is a single stream thats connected and all other old and newly created stream keeps failing which sounds less likely. In this case it will recreate all remaining stream if last non failing stream fails.

Reply:
I don't think that is always true. I've seen scenarios where the ending of streams can be staggered ever so slightly and in some cases `_fillStreamPool` will need to be called multiple times.

Reply:
chance of that scenario could be reduced by decreasing timeout and increasing maxRetries. WDYT of adding call to  _fillStreamPool  in keepAliveHandler if there are not stream failure since specified timeout.

Reply:
I like the idea of improving our refill strategy, but I think putting it in the keep alive handler is a little heavy handed. I think even just wrapping `fillStreamPool` in something like `setImmediate` might do the trick.

Reply:
fillStreamPool is actually a async function and its getting executed in separate event loop than the one that failed stream.

Reply:
Sorry, the thought I was trying to convey is that we might want to slightly stagger the call to fill the pool so that more streams have a chance to close.

Reply:
@callmehiphop Added fix to fill up pool if stream errors out after retries are exhausted and there are less active stream in a pool. Also changed logic for retry count to count multiple stream failure as single retry.